File: ssafy_task/_260206/codeTree_맵방문하기.java
package ssafy_task._260206;

import java.util.*;
import java.io.*;

public class codeTree_맵방문하기 {
    static int R, C, cnt;
    static char[][] board;
    static int[][] visited; // 시간절약을 위한 int visited

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());
        board = new char[R][C];
        visited = new int[R][C];

        for (int i = 0; i < R; i++) {
            board[i] = br.readLine().toCharArray();
        }

        int ans = 0;
        cnt = 0;

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                ans = Math.max(ans, normalBfs(i, j));
            }
        }

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                int tmpInt = bfs(i, j);
                ans = Math.max(ans, tmpInt);
            }
        }
        System.out.println(ans);
    }

    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    static char[] magicKeyword = {'U', 'D', 'L', 'R'};
    
    static int normalBfs(int r, int c) {
        Deque<Integer[]> dq = new ArrayDeque<>();
        dq.offer(new Integer[]{r, c});
        int max = 0;
        cnt++;
        while (!dq.isEmpty()) {
            Integer[] now = dq.pollFirst();

            char nextDir = board[now[0]][now[1]]; // 현재 방향 체크
            int dir = 0;

            if (nextDir == 'U') dir = 0;
            else if(nextDir == 'D') dir = 1;
            else if (nextDir == 'L') dir = 2;
            else if (nextDir == 'R') dir = 3;
            
            int nr = now[0] + dr[dir];
            int nc = now[1] + dc[dir];

            if (nr < 0 || nr >= R || nc < 0 || nc >= C || (visited[nr][nc] == cnt)) continue;

            visited[nr][nc] = cnt;
            dq.offer(new Integer[] {nr, nc});
            max++;
        }
        return max;
    }
    

    static int bfs(int r, int c) {
        int max = 0;
        for (int i = 0; i < R+C+1; i++) { // 마법 부릴 행열번호
            int magic = i<R ? i : i-R;
            for (int j = 0; j < 4; j++) { // 0~3 각각 방향
                int localMax = 1;
                cnt++;

                Deque<Integer[]> dq = new ArrayDeque<>();
                dq.offer(new Integer[]{r, c});
                visited[r][c] = cnt;

                while (!dq.isEmpty()) {
                    Integer[] now = dq.pollFirst();
                    int dir = 0;
                    char nextDir = board[now[0]][now[1]]; // 현재 방향 체크

                    if (i < R && now[0] == magic) {
                        nextDir = magicKeyword[j];
                    } else if (i >= R && now[1] == magic) {
                        nextDir = magicKeyword[j];
                    }

                    if (nextDir == 'U') dir = 0;
                    else if(nextDir == 'D') dir = 1;
                    else if (nextDir == 'L') dir = 2;
                    else if (nextDir == 'R') dir = 3;
                    
                    int nr = now[0] + dr[dir];
                    int nc = now[1] + dc[dir];

                    if (nr < 0 || nr >= R || nc < 0 || nc >= C || (visited[nr][nc] == cnt)) continue;

                    visited[nr][nc] = cnt;
                    dq.offer(new Integer[] {nr, nc});
                    localMax++;
                    }
                    max = Math.max(max, localMax);
                }
            }
            

        return max;
    }
}

File: ssafy_task/_260206/codeTree_맵방문하기.py
import sys
input = sys.stdin.readline

R, C = map(int, input().split())
board = []
visited = [[[0 for _ in range(C)] for _ in range(R)] for _ in range(5)] 
cnt = 0
ans = 0
for i in range(R):
    board.append(list(input().strip()))

dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]
    
def bfs(board, r, c):
    global cnt
    global ans
    cnt += 1
    tot = 1
    if (board[r][c] == 'U'):
        dir = 0
    elif (board[r][c] == 'D'):
        dir = 1
    elif (board[r][c] == 'L'):
        dir = 2
    else:
        dir = 3
    nr = r + dr[dir]
    nc = c + dc[dir]

    while (0 <= nr < R and 0 <= nc < C and visited[dir][nr][nc] != cnt):
        visited[dir][nr][nc] = cnt
        tot += 1
        if (board[nr][nc] == 'U'):
            dir = 0
        elif (board[nr][nc] == 'D'):
            dir = 1
        elif (board[nr][nc] == 'L'):
            dir = 2
        else:
            dir = 3
        nr += dr[dir]
        nc += dc[dir]
    ans = max(ans, tot)

for i in range(R):
    uBoard = [a[:] for a in board]
    uBoard[i] = ['U'] * C
    dBoard = [a[:] for a in board]
    dBoard[i] = ['D'] * C
    lBoard = [a[:] for a in board]
    lBoard[i] = ['L'] * C
    rBoard = [a[:] for a in board]
    rBoard[i] = ['R'] * C
    
    for x in range(R):
        for y in range(C):
            bfs(uBoard, x, y)
            bfs(dBoard, x, y)
            bfs(lBoard, x, y)
            bfs(rBoard, x, y)

for i in range(C):
    uBoard = [a[:] for a in board]
    dBoard = [a[:] for a in board]
    lBoard = [a[:] for a in board]
    rBoard = [a[:] for a in board]
    for j in range(R):
        uBoard[j][i] = 'U'
        dBoard[j][i] = 'D'
        lBoard[j][i] = 'L'
        rBoard[j][i] = 'R'
    
    for x in range(R):
        for y in range(C):
            bfs(uBoard, x, y)
            bfs(dBoard, x, y)
            bfs(lBoard, x, y)
            bfs(rBoard, x, y)

print(ans)